### PR TITLE
Refactor branchprotection config.

### PR DESCRIPTION
**Pull request by fejta**
Take advantage of embedded types.
Prepare to deprecate protection dialect in favor of github's naming.

This should make https://github.com/kubernetes/test-infra/pull/7613 easier to review

/assign @stevekuznetsov @cjwagner

### CODE REVIEW

File: prow/config/config.go [hunk 1]
```diff
@@ -46,12 +47,12 @@ type Config struct {
 	// Periodics are not associated with any repo.
 	Periodics []Periodic `json:"periodics,omitempty"`
 
-	Tide             Tide             `json:"tide,omitempty"`
-	Plank            Plank            `json:"plank,omitempty"`
-	Sinker           Sinker           `json:"sinker,omitempty"`
-	Deck             Deck             `json:"deck,omitempty"`
-	BranchProtection BranchProtection `json:"branch-protection,omitempty"`
```
Comment (cjwagner) [inline prow/config/config.go]: We avoid embedding structs throughout this file and some other parts of Prow where it doesn't make sense for the parent struct to take on the functions of the child struct (if any functions exist).
If we add functions to the BranchProtection or PushGateway structs they shouldn't be exposed from the Config struct in order to maintain encapsulation and avoid conflicting function names.

File: prow/cmd/branchprotector/protect.go [hunk 2]
```diff
@@ -169,13 +168,17 @@ type Protector struct {
 
 func (p *Protector) ConfigureBranches() {
 	for r := range p.updates {
-		if !r.Protect {
+		if r.Policy.Protect == nil {
+			p.errors.add(fmt.Errorf("%s/%s:%s has Protect==nil", r.Org, r.Repo, r.Branch))
```
Comment (stevekuznetsov) [inline prow/cmd/branchprotector/protect.go]: If this is never valid can we enforce it somehow before this caller gets here?

File: prow/cmd/branchprotector/protect.go [hunk 3]
```diff
@@ -188,51 +191,70 @@ func (p *Protector) ConfigureBranches() {
...
-		}
-	}
+	// TODO(fejta): check for deprecated usage and warn
+	policy := bp.MakePolicy()
 
-	// Do not automatically protect tested repositories
-	if !bp.ProtectTested {
-		return
-	}
+	if bp.ProtectTested { // Automatically require any required prow jobs for unconfigured repos
```
Comment (stevekuznetsov) [inline prow/cmd/branchprotector/protect.go]: Shouldn't we also require `ProwJobs` for repos that happen to have something already configured?

File: prow/cmd/branchprotector/protect.go [hunk 4]
```diff
@@ -188,51 +191,70 @@ func (p *Protector) ConfigureBranches() {
...
+		for repoPath := range p.cfg.Presubmits {
+			parts := strings.Split(repoPath, "/")
+			if len(parts) != 2 {
+				p.errors.add(fmt.Errorf("bad repo in prow jobs: %s", repoPath))
+				continue
+			}
+			orgName := parts[0]
+			repoName := parts[1]
+			repoReqs := repoRequirements(orgName, repoName, *p.cfg)
```
Comment (stevekuznetsov) [inline prow/cmd/branchprotector/protect.go]: Can't we partition this by branch by looking at the prow job configs?

File: prow/cmd/branchprotector/protect_test.go [hunk 5]
```diff
@@ -365,317 +393,244 @@ func split(branch string) (string, string, string) {
...
-						case e.Protect != a.Protect:
-							t.Errorf("%s: %s/%s=%s actual protect %t != expected %t", tc.name, e.Org, e.Repo, e.Branch, a.Protect, e.Protect)
-						case len(e.Contexts) != len(a.Contexts), len(e.Contexts) > 0 && !reflect.DeepEqual(e.Contexts, a.Contexts):
-							t.Errorf("%s: %s/%s=%s actual contexts %v != expected %v", tc.name, e.Org, e.Repo, e.Branch, a.Contexts, e.Contexts)
-						case len(e.Pushers) != len(a.Pushers), len(e.Pushers) > 0 && !reflect.DeepEqual(e.Pushers, a.Pushers):
-							t.Errorf("%s: %s/%s=%s actual pushers %v != expected %v", tc.name, e.Org, e.Repo, e.Branch, a.Pushers, e.Pushers)
-						}
-						break
+			var cfg config.Config
+			err := yaml.Unmarshal([]byte(tc.config), &cfg)
```
Comment (stevekuznetsov) [inline prow/cmd/branchprotector/protect_test.go]: Not a fan of this at all, before you have methods that operate on Go structs so you had provided them with the data they needed and checked the output. Now, you added YAML marshaling and unmarshalling to the "unit" under test. If I want to rename or refactor something in my IDE, none of these opaque strings will be changed.

File: prow/config/branchprotection/merge.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+// selectBool returns the child if set, else parent
+func selectBool(parent, child *bool) *bool {
+	if child != nil {
+		return child
+	}
+	return parent
+}
+
+// unionStrings returns a new list after appending both parent and child items.
```
Comment (stevekuznetsov) [inline prow/config/branchprotection/merge.go]: Use k8s `sets.String` ?

File: prow/config/config.go [hunk 7]
```diff
@@ -78,11 +79,11 @@ type Config struct {
 	LogLevel string `json:"log_level,omitempty"`
 
 	// PushGateway is a prometheus push gateway.
-	PushGateway PushGateway `json:"push_gateway,omitempty"`
+	PushGateway `json:"push_gateway,omitempty"`
```
Comment (stevekuznetsov) [inline prow/config/config.go]: These seem like unrelated changes? Why?

File: prow/config/config.go [hunk 8]
```diff
@@ -78,11 +79,11 @@ type Config struct {
...
 
 	// OwnersDirBlacklist is used to configure which directories to ignore when
 	// searching for OWNERS{,_ALIAS} files in a repo.
-	OwnersDirBlacklist OwnersDirBlacklist `json:"owners_dir_blacklist,omitempty"`
+	OwnersDirBlacklist `json:"owners_dir_blacklist,omitempty"`
```
Comment (stevekuznetsov) [inline prow/config/config.go]: These seem like unrelated changes? Why?

File: prow/github/types.go [hunk 9]
```diff
@@ -215,21 +215,32 @@ type Branch struct {
 	Protected bool   `json:"protected"`
 }
 
-// https://developer.github.com/v3/repos/branches/#update-branch-protection
+// BranchProtectionRequest restricts who can write to a branch.
+// All keys required. Disable options by passing a nil value.
+//
+// See: https://developer.github.com/v3/repos/branches/#update-branch-protection
 type BranchProtectionRequest struct {
```
Comment (stevekuznetsov) [inline prow/github/types.go]: Why embed these fields??

File: prow/config/config.go (hunk 1 above)
Comment (fejta) [reply prow/config/config.go]: Done

File: prow/github/types.go (hunk 9 above)
Comment (fejta) [reply prow/github/types.go]: Done

File: prow/config/config.go (hunk 8 above)
Comment (fejta) [reply prow/config/config.go]: Done

File: prow/config/config.go (hunk 7 above)
Comment (fejta) [reply prow/config/config.go]: Done

File: prow/config/branchprotection/merge.go (hunk 6 above)
Comment (fejta) [reply prow/config/branchprotection/merge.go]: Done

File: prow/cmd/branchprotector/protect_test.go (hunk 5 above)
Comment (fejta) [reply prow/cmd/branchprotector/protect_test.go]: This is a better way to write this test. What matters is that a given yaml input produces the same github client calls.

File: prow/cmd/branchprotector/protect.go (hunk 4 above)
Comment (fejta) [reply prow/cmd/branchprotector/protect.go]: Done

**Comment by fejta** [comment]
Moving the piece where we make TestProtect into an integration test which ensures a yaml document is processed correctly into https://github.com/kubernetes/test-infra/pull/7891

Let's start there as this will make this one easier to read.

**Comment by fejta** [comment]
PTAL, these changes are now limited to making the `Policy` refactor, which will make it easier to add more fields in the future.

No prow/cmd/branchprotection tests needed to be changed at this point

File: prow/config/branch_protection.go [hunk 10]
```diff
@@ -22,102 +22,123 @@ import (
...
+		Contexts: unionStrings(p.Contexts, c.Contexts),
+		Pushers:  unionStrings(p.Pushers, c.Pushers),
+	}
+}
+
+func (p PolicyStruct) Get() PolicyStruct {
+	return p
+}
+
+type Policy interface {
```
Comment (cjwagner) [inline prow/config/branch_protection.go]: What is the purpose of this interface? Since `Get` always returns `PolicyStruct` it seems like this interface doesn't provide any advantages over simply using a `PolicyStruct` everywhere. I think the interface and all of the `Get` calls could be removed.

Comment (fejta) [reply prow/config/branch_protection.go]: Done, do you like this better?

**Comment by fejta** [comment]
/assign @BenTheElder 

Can you take a look? Cole is partially available due to kubecon :)

File: prow/config/branch_protection_test.go [hunk 11]
```diff
@@ -18,11 +18,168 @@ package config
...
 
-	"k8s.io/apimachinery/pkg/util/sets"
+var (
+	y   = true
+	n   = false
+	yes = &y
+	no  = &n
 )
 
+func fixup(policy *Policy) {
```
Comment (BenTheElder) [inline prow/config/branch_protection_test.go]: perhaps `func normalize(policy *Policy)` since it seems "fixing" == fixing comparison ? /shrug

Comment (fejta) [reply prow/config/branch_protection_test.go]: Done
